### DISCUSSION TITLE
liquidctl support for RGB keyboard

### DISCUSSION THREAD

**Discussion by mauriciocramos** [liquidctl/liquidctl#561] [Q&A]
Thanks to your amazing work I'm successfully using liquidctl to control pump, fan and RGB of the following devices.

1. [Corsair Hydro H150i Pro XT](https://github.com/liquidctl/liquidctl/blob/main/docs/corsair-platinum-pro-xt-guide.md)
2. [Gigabyte RGB Fusion 2.0 5702 Controller](https://github.com/liquidctl/liquidctl/blob/main/docs/gigabyte-rgb-fusion2-guide.md)

I use this script:
```
# Motherboard RGB
liquidctl --match gigabyte initialize
liquidctl --match gigabyte set sync color fixed $3 --verbose
# AIO liquid cooler
liquidctl --match corsair initialize --pump-mode quiet
liquidctl --match corsair set led color fixed 0000ff
liquidctl --match corsair set fan speed 100
```
As liquidctl also supports RGB of non cooling devices, I would like use it to my RGB keyboard in the same script.

Please help me to acomplish that.  I can collect data and test whatever you need me to.  Although I code in Python, I'm not sure I could develop it myself.

Obviously, liquidctl is (still) not ready for that:
```
mauricio@rig:~$ liquidctl list --verbose

Device #0: Corsair Hydro H150i Pro XT
├── Vendor ID: 0x1b1c
├── Product ID: 0x0c22
├── Release number: 0x0100
├── Bus: hid
├── Address: /dev/hidraw0
└── Driver: HydroPlatinum

Device #1: Gigabyte RGB Fusion 2.0 5702 Controller
├── Vendor ID: 0x048d
├── Product ID: 0x5702
├── Release number: 0x0001
├── Bus: hid
├── Address: /dev/hidraw1
└── Driver: RgbFusion2
```
My RGB keyboard is **Redragon MITRA model K551RGB-1** identified as **Evision RGB keyboard** by lsusb
```
mauricio@rig:~$ sudo lsusb -s 001:007 -v

Bus 001 Device 007: ID 320f:5000 Evision RGB Keyboard
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x320f 
  idProduct          0x5000 
  bcdDevice            1.07
  iManufacturer           1 Evision
  iProduct                2 RGB Keyboard
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x003b
    bNumInterfaces          2
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              400mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      79
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     194
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
Device Qualifier (for other device speed):
  bLength                10
  bDescriptorType         6
  bcdUSB               2.00
  bDeviceClass          255 Vendor Specific Class
  bDeviceSubClass         0 
  bDeviceProtocol       255 
  bMaxPacketSize0        64
  bNumConfigurations      1
Device Status:     0x0000
  (Bus Powered)
```
For what it worths, [OpenRGB supported devices](https://gitlab.com/CalcProgrammer1/OpenRGB/-/wikis/Supported-Devices) includes [Redragon Keyboards (and compatibles)](https://gitlab.com/CalcProgrammer1/OpenRGB/-/wikis/Redragon-K556-Devarajas) containing:
* Packet Structure
* Commands
* Unknown Data Block (0x03/0x04)
* Parameter Data Block (0x05/0x06)
* Modes
* Surmount Mode Colors
* Custom Color Keymap
* etc

**Comment by jonasmalacofilho** [comment]
I'm not sure we're ready to support such devices:

- the only abstraction we have for managing several LEDs is very coarse, the channel; to control a keyboard we would probably need zones, and user-defined ones at that;
- as far as I known, fancier RGB keyboards also allow user-defined effects, which we also don't yet have any way to express.

Additionally, the liquidctl CLI assumes that any effects are implemented by the device. Many devices follow that model, or at least support some effects at the firmware level, independently of having something constantly running on the host.

But I'm not sure that's true for keyboards, given the complexity of effects that I generally see... or are those complex/arbitrary user-defined effects being compiled to some type of bytecode, that then executes on the device?

**Reply by mauriciocramos** [reply]
Hey Jonas how are you doing? I perfectly understand whatever you decide master. :-)

I just would like to use liquidctl to control keyboard RGB because my deep learning batch scripts already use liquidctl to increase pump and fan to maximum and also set all mainboard and cooler lights to blue (max cooling) and, at the end of the job, to set all lights to red (min colling).

I'm confident this keyboard effects are all implemented in hardware, because with no software I'm able to switch over all keyboard effect options.

I would be more than satisfied by having only a CLI command to switch from all blue to all red in my keyboard as well.  I'm a high-level Python developer and don't think I'm able to do that myself, but I'm committed to test.

TL;DR: In case you haven't visited the links I sent before, I put some tables here which seems to address some of your concerns,  I confess it's out of my capacity but I just guess if those could fit in your current architecture.

As suggested by [OpenRGB wiki](https://gitlab.com/CalcProgrammer1/OpenRGB/-/wikis/Redragon-K556-Devarajas):

Packet Structure:

Index | Value | Description
-- | -- | --
0x00 | 0x04 |  
0x01 | 0xCC | Checksum LSB
0x02 | 0xCC | Checksum MSB
0x03 | 0xXX | Command
0x04 | 0xNN | Number of bytes
0x05 | 0xNN | Byte offset LSB
0x06 | 0xNN | Byte offset MSB
0x07 | 0x00 |  
0x08+ | 0xXX | Data bytes

>The command field indicates which data bank to access and whether the command is a read or a write.  The read or write operation uses the data starting at 0x08 in the packet of the given number of bytes and reads/writes it at the given offset.  The checksum simply adds together all of the bytes in the packet excluding the first three.

Commands:

Value | R/W | Description | Total Data Length
-- | -- | -- | --
0x00 |   |   |  
0x01 |   | Begin | 0
0x02 |   | End | 0
0x03 | Read | Unknown Data | 0x2C
0x04 | Write | Unknown Data | 0x2C
0x05 | Read | Parameter Data | 0x8C
0x06 | Write | Parameter Data | 0x8C
0x07:0x0F |   |   |  
0x10 | Read | Custom Color Data |  
0x11 | Write | Custom Color Data |  

Parameter Data Block:

Byte Index | Parameter Bytes | Parameter Description
-- | -- | --
0x00 | 1 | Mode (See Modes table)
0x01 | 1 | Brightness (0-5)
0x02 | 1 | Speed (Slow: 5, Fast: 0)
0x03 | 1 | Direction (R: 0, L: 1)
0x04 | 1 | Random Color Flag (0/1)
0x05 | 1 | Mode Red Color
0x06 | 1 | Mode Green Color
0x07 | 1 | Mode Blue Color
0x08 | 1 |  
0x09 | 1 | Background Red Color
0x0A | 1 | Background Green Color
0x0B | 1 | Background Blue Color
0x0C:0x0E | 1 |  
0x0F | 1 | Polling Rate (See table)
0x10 | 1 |  
0x11 | 1 | Surmount mode color (see table)

Modes:

Effect name | Value | Description
-- | -- | --
Go with the stream | 0x01 | Rainbow Wave
Clouds fly | 0x02 | Rainbow Wave, but stretched out
Winding paths | 0x03 | Color Wheel
The trial of light | 0x04 | Spectrum Cycle
Breathing | 0x05 | Breathing (at least they got this right)
Normally on | 0x06 | Static
Pass without trace | 0x07 | Reactive (single key lights when pressed)
Ripple graff | 0x08 | Reactive with ripple from pressed key
Fast run without trace | 0x09 | Reactive with ripple only on key's row
Swift action | 0x0A | I think Razer calls this Starlight
Flowers blooming | 0x0B | Spectrum Cycle per key, very colorful
Snow winter jasmine | 0x0C | Rainbow Wave, but vertical
Hurricane | 0x0D | Zigzag lines from center of keyboard
Accumulate | 0x0E | Colors come in from outer edges and group in middle
Digital times | 0x0F | A slower version of Swift action (starlight)
Both ways | 0x10 | Visor effect that scans back and forth
Surmount | 0x11 | Seems to fade between hues of selected color
Fast and the Furious | 0x12 | Circular rainbow wave that moves to center
Coastal | 0x14 | Per-key control (did they mean "Custom"?)

Surmount Mode Colors:

Surmount Index | Surmount Color
-- | --
0x00 | Red
0x01 | Yellow
0x02 | Green
0x03 | Blue

Polling Rates:

Polling Rate Index | Polling Rate Setting
-- | --
0x00 | 125Hz
0x01 | 250Hz
0x02 | 500Hz
0x03 | 1000Hz

**Reply by jonasmalacofilho** [reply]
That does seem to simplify things, at least for this specific model.

But without solving the issues I mentioned we would only be able to support a handful of keyboards, and it's a bit  weird to do something like that.

But if someone implements the driver for this keyboard and opens a PR, I might accept it, as long as it is of decent enough quality and doesn't negatively impact other drivers/users.

**Reply by mauriciocramos** [reply]
> That does seem to simplify things, at least for this specific model.
> 
> But without solving the issues I mentioned we would only be able to support a handful of keyboards, and it's a bit weird to do something like that.
>
> But if someone implements the driver for this keyboard and opens a PR, I might accept it, as long as it is of decent enough quality and doesn't negatively impact other drivers/users.

I could give it a try, but have no experience in such kind of software.   Would you mind to give me instructions how to develop this driver?  Maybe some driver template, new driver configuration and testing procedures?

Thanks.

**Reply by jonasmalacofilho** [reply]
Sure, take a look at this comment: https://github.com/liquidctl/liquidctl/issues/560#issuecomment-1385076703
